sos/plugins/crash_report.py

```diff
@@ -0,0 +1,249 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+import os
+import sys
+import time
+from sos.plugins import Plugin, RedHatPlugin, DebianPlugin, UbuntuPlugin
+from sos.utilities import sos_get_command_output, find
+from datetime import datetime
+
+
+class CrashLog(Plugin):
+    """Crash kernel report
+    """
+    plugin_name = "crash_report"
+    profiles = ('system', 'debug')
+
+    dbg_vmlinux_path = '/usr/lib/debug/'
+    vmcore_stnd_path = '/var/crash'
+
+    dump_file_prefix = ''
+    crashed_kernel_name = ''
+    vmlinux_path = ''
+    message_to_log = ''
+    vmcore_file_path = ''
+    crash_script = ""
+    # crash result output file
+    crashed_output_file = '/var/log/os_crash_report.txt'
+
+    # list of crash command
+    crash_cmd_list = [
+                    ("sys", "\"SYSTEM INFORMATION:\""),
+                    ("mach", "\"MACHINE SPECIFIC DATA:\""),
+                    ("mach", "\"MACHINE SPECIFIC DATA:\""),
```

Comment:
1. "mach" and "mach -o" gives different info.
"mach" provides machine specific details.
"mach -o" provides firmware information and only supported on OpenPower.
Hence both commands are needed.
On other arch , mach -o prints error in output file and hence can be identified that log is from other arch .

2. We collect log of various command in a single file one after other.
We append description and then add "output result of command" to identify log, else it will be difficult to identify after seeing log.

Example:
I am pasting partial log of file here. Below is the output of sys command.
We added "SYSTEM INFORMATION:" first to identify that log is from sys command and then output result.
----
SYSTEM INFORMATION:

      KERNEL: 4.8.0-39/vmlinux-4.8.0-39-generic
    DUMPFILE: 4.8.0-39/dump.201705251505  [PARTIAL DUMP]
        CPUS: 160
        DATE: Thu May 25 15:06:14 2017